### PR TITLE
Purely refactors paged/attention into `layers/attention` and make hardware differences more obvious with 1 file per hardware.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: server/text_generation_server/utils/import_utils.py [hunk 1]
```diff
@@ -40,11 +41,12 @@ def get_xpu_free_memory(device, memory_fraction):
     synchronize = torch.xpu.synchronize
     get_free_memory = get_xpu_free_memory
 else:
-    SYSTEM = "cpu"
+    SYSTEM = "ipex"
```
Comment (Narsil) [inline server/text_generation_server/utils/import_utils.py:44]:
```suggestion
    SYSTEM = "cpu"
```

File: server/text_generation_server/layers/attention/rocm.py [hunk 2]
```diff
@@ -1,96 +1,169 @@
...
-from text_generation_server.utils.import_utils import SYSTEM
+major, minor = torch.cuda.get_device_capability()
+is_sm75 = major == 7 and minor == 5
+_PARTITION_SIZE = 512
 
-if SYSTEM != "xpu":
-    from text_generation_server.utils.flash_attn_triton import triton_attention
+use_triton = os.getenv("ROCM_USE_FLASH_ATTN_V2_TRITON", "").lower() in {"true", "1"}
+ENGINE = "triton" if use_triton else "ck"
+from .flash_attn_triton import triton_attention
```
Comment (danieldk) [inline server/text_generation_server/layers/attention/rocm.py:12]: Does it make sense to import this when `use_triton == False`?

File: server/text_generation_server/layers/attention/rocm.py [hunk 3]
```diff
@@ -1,96 +1,169 @@
...
+            "auto",
+            1.0,
+        )
+
+
+try:
+    import flash_attn_2_cuda
+
+    if ENGINE == "triton":
+        logger.info("ROCm: using Flash Attention 2 Triton implementation.")
```
Comment (danieldk) [inline server/text_generation_server/layers/attention/rocm.py:129]: In the conditions below, it seems like `flash_attn_2_cuda` is never used when `ENGINE == "triton"`? Also, maybe this only needs to be imported when `ENGINE == "ck"`?

File: server/text_generation_server/layers/attention/xpu.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    v,
+    out,
+    cu_seqlens,
+    max_s,
+    softmax_scale,
+    window_size_left=-1,
+):
+    if window_size_left <= 0 and window_size_left != -1:
+        raise ValueError("`window_size_left` must be > 0 or -1")
+
```
Comment (danieldk) [inline server/text_generation_server/layers/attention/xpu.py:19]: I guess this condition can be removed as well, given that window attention is not supported.

File: server/text_generation_server/layers/attention/rocm.py [hunk 5]
```diff
@@ -101,46 +174,14 @@ def attention(
...
         window_size_left=-1,
+        causal=True,
     ):
         if window_size_left <= 0 and window_size_left != -1:
             raise ValueError("`window_size_left` must be > 0 or -1")
-
         if window_size_left != -1:
             raise ValueError(
-                f"XPU version of Flash Attention does not support window attention (window_size_left != -1, got window_size_left={window_size_left})."
+                f"RoCm version of Flash Attention v2 does not support window attention (window_size_left != -1, got window_size_left={window_size_left})."
```
Comment (danieldk) [inline server/text_generation_server/layers/attention/rocm.py:183]: NIT

```suggestion
                f"ROCm version of Flash Attention v2 does not support window attention (window_size_left != -1, got window_size_left={window_size_left})."
```

File: server/text_generation_server/layers/attention/rocm.py [hunk 6]
```diff
@@ -1,96 +1,169 @@
...
-                if "MI210" not in torch.cuda.get_device_name(
-                    idx
-                ) and "MI250" not in torch.cuda.get_device_name(idx):
+                name = torch.cuda.get_device_name(idx)
+                if "MI210" not in name and "MI250" not in name:
                     raise ImportError(
                         f"AMD GPU {torch.cuda.get_device_name(idx)} does not support flash-attention"
                     )
+            raise ImportError(
+                f"AMD GPU with Rocm capability {major} {minor} is not supported"
```
Comment (danieldk) [inline server/text_generation_server/layers/attention/rocm.py:161]: NIT

```suggestion
                f"AMD GPU with ROCm capability {major} {minor} is not supported"
```

**Review by danieldk — COMMENTED** [review]
The new structure is really much cleaner, left some small comments/questions.

File: server/text_generation_server/layers/attention/rocm.py (hunk 2 above)
Comment (Narsil) [reply server/text_generation_server/layers/attention/rocm.py:12]: Not really. I'll move it behind itś flag.

File: server/text_generation_server/layers/attention/rocm.py (hunk 3 above)
Comment (Narsil) [reply server/text_generation_server/layers/attention/rocm.py:129]: Indeed.

File: server/text_generation_server/layers/attention/rocm.py (hunk 5 above)
Comment (Narsil) [reply server/text_generation_server/layers/attention/rocm.py]: Thanks !

File: server/text_generation_server/layers/attention/rocm.py (hunk 6 above)
Comment (Narsil) [reply server/text_generation_server/layers/attention/rocm.py]: Thanks !

File: server/text_generation_server/layers/attention/xpu.py (hunk 4 above)
Comment (danieldk) [reply server/text_generation_server/layers/attention/xpu.py:19]: But below an exception is raised `window_size_left != -1`, so all window sizes other than `-1` are impossible.

**Review by danieldk — DISMISSED** [review]
Added another nitpick question, LGTM otherwise.

**Pull request merged**